flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/ddl/SqlCompilePlan.java

```diff
@@ -0,0 +1,108 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.sql.parser.ddl;
+
+import org.apache.flink.sql.parser.dml.RichSqlInsert;
+import org.apache.flink.sql.parser.dml.SqlStatementSet;
+
+import org.apache.calcite.sql.SqlCall;
+import org.apache.calcite.sql.SqlCharStringLiteral;
+import org.apache.calcite.sql.SqlKind;
+import org.apache.calcite.sql.SqlNode;
+import org.apache.calcite.sql.SqlOperator;
+import org.apache.calcite.sql.SqlSpecialOperator;
+import org.apache.calcite.sql.SqlWriter;
+import org.apache.calcite.sql.parser.SqlParserPos;
+import org.apache.calcite.util.NlsString;
+
+import javax.annotation.Nonnull;
+
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * AST node for {@code COMPILE PLAN 'planfile' [IF NOT EXISTS] FOR [DML]}. DML can be either a
+ * {@link RichSqlInsert} or a {@link SqlStatementSet}.
+ */
+public class SqlCompilePlan extends SqlCall {
+
+    public static final SqlSpecialOperator OPERATOR =
+            new SqlSpecialOperator("CompilePlan", SqlKind.OTHER);
+
+    private final SqlCharStringLiteral planFile;
+    private final boolean ifNotExists;
+    private SqlNode operand;
+
+    public SqlCompilePlan(
+            SqlParserPos pos, SqlCharStringLiteral planFile, boolean ifNotExists, SqlNode operand) {
+        super(pos);
+        this.planFile = planFile;
+        this.ifNotExists = ifNotExists;
+        this.operand = checkOperand(operand);
+    }
+
+    public String getPlanFile() {
+        return planFile.getValueAs(NlsString.class).getValue();
+    }
+
+    public boolean isIfNotExists() {
+        return ifNotExists;
+    }
+
+    @Nonnull
+    @Override
+    public SqlOperator getOperator() {
+        return OPERATOR;
+    }
+
+    @Nonnull
+    @Override
+    public List<SqlNode> getOperandList() {
+        return Collections.singletonList(operand);
+    }
+
+    @Override
+    public void unparse(SqlWriter writer, int leftPrec, int rightPrec) {
+        writer.keyword("COMPILE PLAN");
+        planFile.unparse(writer, leftPrec, rightPrec);
+        if (isIfNotExists()) {
+            writer.keyword("IF NOT EXISTS");
+        }
+        writer.keyword("FOR");
+        operand.unparse(writer, leftPrec, rightPrec);
+    }
+
+    @Override
+    public void setOperand(int i, SqlNode operand) {
+        if (i == 0) {
+            this.operand = checkOperand(operand);
+        } else {
+            throw new UnsupportedOperationException(
+                    "SqlCompilePlan supports only one operand with index 0");
+        }
+    }
+
+    private SqlNode checkOperand(SqlNode operand) {
+        if (!(operand instanceof RichSqlInsert || operand instanceof SqlStatementSet)) {
+            throw new UnsupportedOperationException(
+                    "SqlCompilePlan supports only RichSqlInsert or SqlStatementSet as operand");
```

Comment:
Could a user see this? If yes, remove internal class names and write the exception like: `COMPILE PLAN supports only INSERT INTO or STATEMENT SET definitions.`

Reply:
No this happens only in case of a parser bug or a validator bug. It would probably be better to use the `assert` statement instead of `UnsupportedOperationException`, but the reason i used it is for consistency.